Jenkinsfile

```diff
@@ -1,15 +1,34 @@
 #!/usr/bin/groovy
 node{
   ws{
+
     stage 'checkout code'
-    checkout scm
+        checkout scm
+
+    stage 'build and unit test'
+    
+        kubernetes.pod('buildtestpod').withImage('maven')
+        .withPrivileged(true)
+        .withSecret('jenkins-maven-settings','/root/.m2')
+        .inside {
+            sh 'mvn -Dmaven.test.failure.ignore clean verify'
+            step([$class: 'JUnitResultArchiver', testResults: '**/target/surefire-reports/TEST-*.xml'])
+        }
 
     stage 'build and deploy'
-    kubernetes.pod('buildpod').withImage('maven')
-    .withPrivileged(true)
-    .withSecret('jenkins-maven-settings','/root/.m2')
-    .inside {
-        sh 'mvn clean deploy'
-    }
+        kubernetes.pod('deploypod').withImage('maven')
+        .withPrivileged(true)
+        .withSecret('jenkins-maven-settings','/root/.m2')
+        .inside {
+            try {
+                sh 'mvn clean deploy'
```

Comment:
* There's no Nexus :)  To save in overhead we're going to expose the local Maven repository over a slim HTTP server.
* Cool, let's kill this clean
* Understood now on your last point - we need to force a non-0 exit to fail the build and we can't get that without re-running the tests (can't get it from just the test results XMLs unfortunately)